### PR TITLE
Move I/O buffers and control to data coordinator

### PR DESCRIPTION
**Pull request by bvanessen**
Created a buffered data coordinator child class to hold data type
specific I/O buffers in the data coordinator.  This moves the
functionality for holding the data out of the input layer.

### CODE REVIEW

**Comment by bvanessen** [comment]
@timmoon10 @benson31 I believe that this PR is ready for review.  It is passing all bamboo tests locally except for tessellate.   When reviewing this please remember that this is supposed to be as close to a "straight" refactor as possible.  There are a number of future optimizations that we will want to do, but I wanted to get this pushed out first before adding more changes.

File: src/io/CMakeLists.txt [hunk 1]
```diff
@@ -1,4 +1,4 @@
-add_subdirectory(data_buffers)
+#add_subdirectory(data_buffers)
```
Comment (timmoon10) [inline src/io/CMakeLists.txt:1]:
```suggestion
```

File: src/training_algorithms/sgd_training_algorithm.cpp [hunk 2]
```diff
@@ -138,7 +144,7 @@ bool sgd_training_algorithm::train_mini_batch(sgd_execution_context& c,
 
   // Update step
   model.update_weights();
-  finished = model.update_layers();
+  /*finished = */model.update_layers();
```
Comment (timmoon10) [inline src/training_algorithms/sgd_training_algorithm.cpp:147]:
```suggestion
  model.update_layers();
```

**Review by timmoon10 — APPROVED** [review]
LGTM

**Pull request merged**